cc/main.cc

```diff
@@ -282,36 +319,72 @@ std::thread SelfPlayThread(int thread_id, ThreadSafeRandom* rnd,
       auto output_name = GetOutputName(now, thread_id);
 
-      bool is_holdout = (*rnd)() < FLAGS_holdout_pct;
-      auto example_dir = is_holdout ? FLAGS_holdout_dir : FLAGS_output_dir;
+      bool is_holdout;
+      {
+        absl::MutexLock lock(&mutex_);
+        is_holdout = rnd_() < game_options.holdout_pct;
+      }
+      auto example_dir =
+          is_holdout ? game_options.holdout_dir : game_options.output_dir;
       if (!example_dir.empty()) {
         WriteExample(GetOutputDir(now, example_dir), output_name, *player);
       }
 
-      if (!FLAGS_sgf_dir.empty()) {
-        WriteSgf(GetOutputDir(now, file::JoinPath(FLAGS_sgf_dir, "clean")),
-                 output_name, *player, false);
-        WriteSgf(GetOutputDir(now, file::JoinPath(FLAGS_sgf_dir, "full")),
-                 output_name, *player, true);
+      if (!game_options.sgf_dir.empty()) {
+        WriteSgf(
+            GetOutputDir(now, file::JoinPath(game_options.sgf_dir, "clean")),
+            output_name, *player, false);
+        WriteSgf(
+            GetOutputDir(now, file::JoinPath(game_options.sgf_dir, "full")),
+            output_name, *player, true);
       }
-    } while (run_forever);
-  });
-}
-
-void SelfPlay() {
-  auto dual_net_factory = NewDualNetFactory(FLAGS_model);
-  MctsPlayer::Options options;
-  ParseMctsPlayerOptionsFromFlags(&options);
-  ThreadSafeRandom rnd;
+    } while (game_options.run_forever);
 
-  std::vector<std::thread> threads;
-  for (int i = 0; i < FLAGS_parallel_games; ++i) {
-    threads.push_back(SelfPlayThread(i, &rnd, options, dual_net_factory.get(),
-                                     FLAGS_run_forever));
+    std::cerr << "Thread " << thread_id << " stopping" << std::endl;
   }
-  for (auto& t : threads) {
-    t.join();
+
+  void MaybeReloadFlags() EXCLUSIVE_LOCKS_REQUIRED(&mutex_) {
+    if (FLAGS_flags_path.empty()) {
+      return;
+    }
+    uint64_t new_flags_timestamp;
+    TF_CHECK_OK(tf_utils::GetModTime(FLAGS_flags_path, &new_flags_timestamp));
+    if (new_flags_timestamp == flags_timestamp_) {
+      return;
+    }
+
+    flags_timestamp_ = new_flags_timestamp;
+    std::string contents;
+    TF_CHECK_OK(tf_utils::ReadFile(FLAGS_flags_path, &contents));
+
+    for (auto line : absl::StrSplit(contents, '\n')) {
+      std::pair<absl::string_view, absl::string_view> line_comment =
+          absl::StrSplit(line, absl::MaxSplits('#', 1));
+      line = absl::StripAsciiWhitespace(line_comment.first);
+      if (line.empty()) {
+        continue;
+      }
+      MG_CHECK(line.length() > 2 && line[0] == '-' && line[1] == '-') << line;
+      std::pair<std::string, std::string> flag_value =
+          absl::StrSplit(line, absl::MaxSplits('=', 1));
+      flag_value.first = flag_value.first.substr(2);
+      std::cerr << "Setting command line flag: --" << flag_value.first << "="
+                << flag_value.second << std::endl;
+      gflags::SetCommandLineOption(flag_value.first.c_str(),
```

Comment:
Not even sure if this is their API but Can you use ProcessOptionsFromStringLocked?
https://github.com/gflags/gflags/blob/23348465a02c9fdbdecce01cee24e62b9f1f5bad/src/gflags.cc#L968

Reply:
Or maybe even ProcessFlagfileLocked

Reply:
Neither of those is part of the gflags public API.

Reply:
Cloud you use ParseCommandLineFlags and pass the array of lines from the file as argv?

There's also ReadFromFlagsFile but it's deprecated.

Reply:
Kinda hacky but you can just set argv = ["--flagfile", "<COPY_OF_GCS_FILE>"] and call ParseCommandLineFlags

Suggested by author here
https://github.com/gflags/gflags/issues/231#issuecomment-341087615

Reply:
ReadFromFlagsFile was apparently marked deprecated 12 years ago, so I doubt it's going away any time soon. Still, we should probably be good citizens and not use it.

I guess ParseCommandLineFlags would work, but I'm not sure it's much better than what I have.